### ISSUE TITLE
wwm assumes user's task bar is at bottom of screen

### ISSUE THREAD

**Issue by ramirezmike** [TimUntersberger/wwm#96]
When the user has wwm setup to display the windows taskbar, it shifts all managed tiles up by the taskbar size. But, if the user happens to keep their taskbar at the top, left or right side of the screen it still behaves as if the taskbar is at the bottom of the screen.

This causes overlap of the taskbar over tiles and introduces an unusable gap at the bottom of the screen.

I've started looking at it and have some code for detecting where the taskbar is and handling it accordingly just for my personal use (I tend to keep my taskbar at the top of the screen) but I still have to code up handling left/right placement.

The following things should probably be kept in mind:

1) detecting and handling when the user moves the taskbar while wwm is running
2) how to handle when users have taskbars displaying on multiple monitors (and multiple positions?) I usually just keep the taskbar on my primary monitor, but some users may prefer having it display on every monitor
3) keeping the code flexible enough to not make it a pain to implement a feature that lets the user set the position of the wwm appbar to somewhere other than the top of the screen (a separate issue to write up?)

I can keep working on this and submit a pull request at some point while staying compatible with the development, resize and modes branches if that's helpful.

**Comment by TimUntersberger** [comment]
I will assign you this issue if you want to work on this.

Being able to position the appbar should probably be a seperate issue.

My first instincts regarding the taskbar data is to save it somehow in each `Display` instance instead of globally. If possible could you push your current work to your fork or a new branch on this repo so we can talk about your solution?

**Comment by ramirezmike** [comment]
>My first instincts regarding the taskbar data is to save it somehow in each Display instance instead of globally

Yeah, definitely. I'm working on consolidating some of the taskbar/appbar logic that alters the display size so that it behaves more like calling the width() and height() functions gives you the current "working area" of the current display with respect to what's setup on it.

I'll push to my fork when I get a chance and we can discuss it more.

**Comment by ramirezmike** [comment]
I had some time to look at this again and made some good progress. It's late here so I'm probably going to finish cleaning up and uploading something to my fork tomorrow morning, but I've got it so it's working pretty well.

I've pretty much _emptied_ src/task_bar.rs into src/display.rs to make each display own its own reference to its taskbar if it exists and then updated all references to show/hide to instead work against the display collection showing/hiding the taskbar. I've also updated the code for width()/height() on display to factor in whether or not that display's taskbar is showing and how it affects it so that the appbar and tile grid plays nice with whatever state a given display is in.

There are a couple things I still want to address:

A) Detecting when the user changes the taskbar and updating it. Like, either moving it or removing it through Windows settings.
B) There are a few scenarios where the appbar doesn't update after show/hide'ing the taskbar, but I think that should be pretty easy to resolve.

Here are a couple pictures of different positions

Taskbar on left side:

![Screenshot from 2020-08-08 23-33-16](https://user-images.githubusercontent.com/1421719/89724408-c1702a80-d9d0-11ea-8ec3-f0cb32f68c59.png)

From above picture, stretching out the taskbar:

![Screenshot from 2020-08-08 23-33-39](https://user-images.githubusercontent.com/1421719/89724418-d220a080-d9d0-11ea-9163-30a015f10a66.png)

Just to show what I'm still working on, here I've placed the taskbar on the right side and then moved it to the top and you can see it introduces a gap on the right and there is overlap between everything along the top

![Screenshot from 2020-08-08 23-34-43](https://user-images.githubusercontent.com/1421719/89724451-0eec9780-d9d1-11ea-9f3d-1d503cb31042.png)

But, just using the toggle_config bindings to hide both the taskbar and the appbar and re-showing them results in it displaying like the following:

![Screenshot from 2020-08-08 23-34-59](https://user-images.githubusercontent.com/1421719/89724460-20ce3a80-d9d1-11ea-88d9-23e7c75f6d18.png)


And this is all on latest development branch

**Comment by TimUntersberger** [comment]
Nice work!

Please ping me when you pushed your code.

**Comment by ramirezmike** [comment]
@TimUntersberger I pushed to my fork's development branch you can see [here](https://github.com/ramirezmike/wwm/commit/70ae4da6eca6ea142df679a8145f2f939a0c9baa).

A couple things to keep in mind:

1.  I'm iterating over the displays and hiding/showing the taskbar and it works on my primary display but for some reason it's just.. hiding and then re-showing the taskbar on my secondary display. This is something I'm still looking into unfortunately.
2.  I moved a lot of taskbar code into display.rs while I was working on it but now that it's "done" I can see either pulling stuff back into task_bar.rs or leaving it in display.rs
3.  This doesn't handle if the user moves the position of the taskbar to another edge (top, bottom, left or right) automatically but the workaround is just showing/hiding the taskbar/appbar and then everything shows up correctly. I figure it's something that can be handled in a separate issue

At a high level, I moved taskbar code into a struct that each display owns. When initializing displays it enumerates windows and checks if a given window is a taskbar that is on the current display and if it is it saves information for it.

**Comment by ramirezmike** [comment]
There's something I'm not quite getting with the show/hide of the taskbar... It's as if calling hide on one shows the other. The loop I have in the checkin calls ShowWindow with SW_HIDE for the taskbar on each display and it hides the secondary one first then hides the primary... but then the secondary one gets shown again. 

If I setup the loop to only call ShowWindow once per collection then I end up with the opposite where it hides the taskbar on the secondary and leaves it visible on the primary... Not sure if it's a strange Windows thing I'm unaware of..

**Comment by TimUntersberger** [comment]
> There's something I'm not quite getting with the show/hide of the taskbar... It's as if calling hide on one shows the other. The loop I have in the checkin calls ShowWindow with SW_HIDE for the taskbar on each display and it hides the secondary one first then hides the primary... but then the secondary one gets shown again.

I fixed this before. If you merge the latest changes from `development` you should have the fix.

**Comment by TimUntersberger** [comment]
Can you open a PR so I can easily comment on code snippets? Just title it `WIP: ....`

**Comment by ramirezmike** [comment]
I added a pull request but I think I'd need to recommit to change the name. I haven't made any changes yet so it's still a WIP. It should be close to latest on development branch.

**Comment by ramirezmike** [comment]
I've updated based on comments. The way it works now, if the user does alter taskbar/display settings for some reason, they can use their configured key commands for toggling taskbar/appbar to make everything appear correctly. In the future some display event detection, perhaps as part of #90 , can handle those scenarios and the toggle of the taskbar could be simplified to just show/hide and not needing to update the displays' references.
